### PR TITLE
router: per-cluster histograms w/ timeout budget

### LINKED ISSUE #6122: add per-cluster histograms with remaining timeout budget
Since timeouts are dynamic (e.g. controllable via header values), it is difficult to observe how close requests are to hitting the timeout in production.

A histogram of remaining timeout budget would allow the user to identify latencies that are trending towards the timeout cliff as well as identify poorly tuned timeouts.

Likely a percentage histogram would be the most useful stat, i.e. p95 of requests to upstream are within 80% of their timeout budget.

Also see [#6121: add computed-timeout header](https://github.com/envoyproxy/envoy/issues/6121).

### PR DESCRIPTION
**Pull request by gumpt**
This change adds a pair of histograms to track timeout budget usage: one for the
per-request timeout and one for the global timeout.  The histograms are scaled
such that 100% of the timeout budget used (a timeout) is a value of 100.

The histograms by default are off.  They can be turned on by setting
`track_timeout_budgets` on a cluster to true.

Risk Level: Low (new stats only).
Testing: Unit tests added/modified.
Docs Changes: Added notes about histograms.
Release Notes: Added comment with link to docs.

Fixes #6122

Signed-off-by: Matthew Gumport <mgumport@lyft.com>

### CODE REVIEW

File: test/common/router/router_test.cc [hunk 1]
```diff
@@ -1115,6 +1116,77 @@ TEST_F(RouterTest, UpstreamTimeout) {
...
+
+  Http::TestHeaderMapImpl headers{{"x-envoy-upstream-rq-timeout-ms", "200"},
+                                  {"x-envoy-upstream-rq-per-try-timeout-ms", "100"}};
+  HttpTestUtility::addDefaultHeaders(headers);
+  router_.decodeHeaders(headers, false);
+  Buffer::OwnedImpl data;
+  router_.decodeData(data, true);
+
+  EXPECT_CALL(cm_.thread_local_cluster_.cluster_.info_->stats_store_,
+              deliverHistogramToSinks(_, 50ull))
```
Comment (tonya11en) [inline test/common/router/router_test.cc]: Is this your new histogram or something else?

File: test/mocks/stats/mocks.h [hunk 2]
```diff
@@ -318,6 +318,9 @@ class MockIsolatedStatsStore : public SymbolTableProvider, public IsolatedStoreI
   ~MockIsolatedStatsStore() override;
 
   MOCK_METHOD2(deliverHistogramToSinks, void(const Histogram& histogram, uint64_t value));
+  MOCK_METHOD2(histogram, Histogram&(const std::string&, Histogram::Unit));
```
Comment (tonya11en) [inline test/mocks/stats/mocks.h]: Is this method used anywhere? I might just be missing something, but it's not clear to me why it's necessary.

File: docs/root/configuration/upstream/cluster_manager/cluster_stats.rst [hunk 3]
```diff
@@ -69,7 +69,9 @@ Every cluster has a statistics tree rooted at *cluster.<name>.* with the followi
   upstream_rq_cancelled, Counter, Total requests cancelled before obtaining a connection pool connection
   upstream_rq_maintenance_mode, Counter, Total requests that resulted in an immediate 503 due to :ref:`maintenance mode<config_http_filters_router_runtime_maintenance_mode>`
   upstream_rq_timeout, Counter, Total requests that timed out waiting for a response
+  upstream_rq_timeout_budget_global, Histogram, What percentage of the global timeout was used waiting for a response
```
Comment (tonya11en) [inline docs/root/configuration/upstream/cluster_manager/cluster_stats.rst]: Can you mention the idiosyncrasies around what happens when there is no timeout configured and what default values show up  when?

File: test/mocks/stats/mocks.h (hunk 2 above)
Comment (gumpt) [reply test/mocks/stats/mocks.h]: Remnants from trying to be more clever than the testing approach allowed from this class, removed.

File: docs/root/configuration/upstream/cluster_manager/cluster_stats.rst (hunk 3 above)
Comment (gumpt) [reply docs/root/configuration/upstream/cluster_manager/cluster_stats.rst]: If the timeout is set to 0/infinite, then the histogram will be filled with zeroes.  Otherwise, the time used out of the default timeout will be written.

Are you suggesting this be documented explicitly in this file?

File: source/common/router/router.cc [hunk 4]
```diff
@@ -90,6 +90,21 @@ bool convertRequestHeadersForInternalRedirect(Http::HeaderMap& downstream_header
...
+// values, and getting multiple significant figures on the histogram would be nice.
+uint64_t percentageOfTimeout(const std::chrono::milliseconds response_time,
+                             const std::chrono::milliseconds timeout) {
+  // Timeouts of 0 are considered infinite. Any portion of an infinite timeout used is still
+  // none of it.
+  if (timeout.count() == 0) {
+    return 0;
+  }
+
+  return static_cast<uint64_t>(response_time.count() * TimeoutPrecisionFactor / timeout.count());
```
Comment (tonya11en) [inline source/common/router/router.cc]: nit: I think this static_cast is unnecessary since the function already returns a uint64_t.

File: test/common/router/router_test.cc [hunk 5]
```diff
@@ -1115,6 +1116,81 @@ TEST_F(RouterTest, UpstreamTimeout) {
...
+  Buffer::OwnedImpl data;
+  router_.decodeData(data, true);
+
+  // These are coming from CodeStatsImpl::chargeResponseTiming.
+  EXPECT_CALL(cm_.thread_local_cluster_.cluster_.info_->stats_store_,
+              deliverHistogramToSinks(_, 80ull))
+      .Times(3);
+  // Global timeout budget used.
+  EXPECT_CALL(cm_.thread_local_cluster_.cluster_.info_->stats_store_,
+              deliverHistogramToSinks(_, 20ull));
```
Comment (tonya11en) [inline test/common/router/router_test.cc]: Can you be explicit about which histogram is getting which value in these tests. I think you can do something like:
https://github.com/envoyproxy/envoy/blob/master/test/extensions/filters/network/redis_proxy/command_splitter_impl_test.cc#L602

File: test/integration/stats_integration_test.cc [hunk 6]
```diff
@@ -259,7 +259,8 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithFakeSymbolTable) {
   // 2019/10/17  8484     43340       44000   stats: add unit support to histogram
   // 2019/11/01  8859     43563       44000   build: switch to libc++ by default
   // 2019/11/15  9040     43371       44000   build: update protobuf to 3.10.1
-  // 2019/11/15  9040     43403       35500   upstream: track whether cluster is local
+  // 2019/11/15  9031     43403       44000   upstream: track whether cluster is local
+  // 2019/12/04  9227     49531       50000   router: per-cluster histograms w/ timeout budget
```
Comment (tonya11en) [inline test/integration/stats_integration_test.cc]: I think we should make these histograms opt-in similar to the way the circuit_breaker.proto has a flag for extra stats:
https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/cluster/circuit_breaker.proto#L49

It would cut down on the default memory usage and I'd argue this stat isn't for everyone. See if you can mimic something like this below?
https://github.com/envoyproxy/envoy/blob/81d3bebf9187f408c829dde367214069651a3768/source/common/upstream/upstream_impl.cc#L1043

File: source/common/router/router.cc (hunk 4 above)
Comment (gumpt) [reply source/common/router/router.cc]: It's there for annotation that there is intentional narrowing in the absence of a `gsl::narrow_cast` type call.

File: docs/root/configuration/upstream/cluster_manager/cluster_stats.rst (hunk 3 above)
Comment (gumpt) [reply docs/root/configuration/upstream/cluster_manager/cluster_stats.rst]: Previous comment is outdated.

Based on later feedback, this won't be set without the correct flag.  Should this be in its own table for clarity of documentation?

File: include/envoy/upstream/upstream.h [hunk 7]
```diff
@@ -614,6 +612,13 @@ class PrioritySet {
   REMAINING_GAUGE(remaining_retries, Accumulate)                                                   \
   REMAINING_GAUGE(remaining_rq, Accumulate)
 
+/**
+ * All stats around timeout budgets. Not used by default.
```
Comment (tonya11en) [inline include/envoy/upstream/upstream.h]: nit: Maybe mention which configuration option enables these?

**Comment by tonya11en** [comment]
Also, can you write a bit about the flag you added that toggles the histograms in the description.

File: api/envoy/api/v2/cds.proto [hunk 8]
```diff
@@ -804,6 +804,11 @@ message Cluster {
   // maybe by allowing LRS to go on the ADS stream, or maybe by moving some of the negotiation
   // from the LRS stream here.]
   core.ConfigSource lrs_server = 42;
+
+  // If track_timeout_budgets is true, the timeout budget histograms will be published for each
```
Comment (mattklein123) [inline api/envoy/api/v2/cds.proto]: Can you describe which stats are emitted and ref-link to those stats. Also describe the timeout precision comment so the user understands what is emitted.

File: docs/root/configuration/upstream/cluster_manager/cluster_stats.rst [hunk 9]
```diff
@@ -69,7 +69,9 @@ Every cluster has a statistics tree rooted at *cluster.<name>.* with the followi
   upstream_rq_cancelled, Counter, Total requests cancelled before obtaining a connection pool connection
   upstream_rq_maintenance_mode, Counter, Total requests that resulted in an immediate 503 due to :ref:`maintenance mode<config_http_filters_router_runtime_maintenance_mode>`
   upstream_rq_timeout, Counter, Total requests that timed out waiting for a response
+  upstream_rq_timeout_budget_percent_used, Histogram, What percentage of the global timeout was used waiting for a response
```
Comment (mattklein123) [inline docs/root/configuration/upstream/cluster_manager/cluster_stats.rst]: Perhaps make a new section for these 2 stats and describe the timeout precision concept.

File: source/common/router/router.cc [hunk 10]
```diff
@@ -737,14 +752,18 @@ void Filter::onResponseTimeout() {
...
       cluster_->stats().upstream_rq_timeout_.inc();
       if (upstream_request->upstream_host_) {
         upstream_request->upstream_host_->stats().rq_timeout_.inc();
       }
 
+      // Record the max-value into the histogram to say that the entire timeout
+      // budget was used.
+      cluster_->timeoutBudgetStats().upstream_rq_timeout_budget_per_try_percent_used_.recordValue(
+          TimeoutPrecisionFactor);
+
```
Comment (mattklein123) [inline source/common/router/router.cc]: Is this always true? The global timeout might occur in the middle of a per-try timeout?

File: include/envoy/upstream/upstream.h [hunk 11]
```diff
@@ -812,6 +826,11 @@ class ClusterInfo {
    */
   virtual ClusterLoadReportStats& loadReportStats() const PURE;
 
+  /**
+   * @return ClusterTimeoutBudgetStats& stats on timeout budgets for this cluster.
+   */
+  virtual ClusterTimeoutBudgetStats& timeoutBudgetStats() const PURE;
```
Comment (mattklein123) [inline include/envoy/upstream/upstream.h]: If the feature is not being used, can we not allocate anything for it? So perhaps return an optional std::reference? If we do that we can:
1) Remove all the null histogram changes I think
2) Make the router code branch and avoid all of the timing code if we are not using this feature?

**Comment by gumpt** [comment]
Need to merge master and get rid of the scope I think, not quite done yet.

File: test/common/router/router_test.cc [hunk 12]
```diff
@@ -1122,6 +1122,164 @@ TEST_F(RouterTest, UpstreamTimeout) {
...
+  expectPerTryTimerCreate();
+  router_.retry_state_->callback_();
+
+  // Global timeout used overall.
+  EXPECT_CALL(
+      cm_.thread_local_cluster_.cluster_.info_->timeout_budget_stats_store_,
+      deliverHistogramToSinks(
+          Property(&Stats::Metric::name, "upstream_rq_timeout_budget_percent_used"), 100ull));
+  // Per-try budget used on the second request before the global timeout hit.
+  EXPECT_CALL(cm_.thread_local_cluster_.cluster_.info_->timeout_budget_stats_store_,
```
Comment (tonya11en) [inline test/common/router/router_test.cc]: Recording our offline conversation for posterity:

In the case where the global timeout is firing in the middle of the per-try timeout, it doesn't make sense to record a value for the per-try. Technically, there was some percentage of the per-try timeout used, but it had nothing to do with the per-try timeout.

In my opinion (and it sounded like you agree with me), we should simply cancel the recording of any per-try budget stat when handling the global timeout. Otherwise, the information is misleading.

File: api/envoy/api/v2/cds.proto [hunk 13]
```diff
@@ -801,6 +801,15 @@ message Cluster {
...
   // from the LRS stream here.]
   core.ConfigSource lrs_server = 42;
+
+  // If track_timeout_budgets is true, the :ref:`timeout budget histograms
+  // <config_cluster_manager_cluster_stats_timeout_budgets>` will be published for each
+  // request. These show what percentage of a request's per try and global timeout was used. A value
+  // of 0 would indicate that none of the timeout was used or that the timeout was infinite. A value
+  // of 100 would indicate that the request took the entirety of the timeout given to it.
+  //
+  // If not specified, the default is false.
```
Comment (htuch) [inline api/envoy/api/v2/cds.proto]: Nit: this is implied by the type, so I don't think it's necessary to call out.

**Comment by gumpt** [comment]
Looks like I merged master while it was red.

File: docs/root/intro/version_history.rst [hunk 14]
```diff
@@ -27,12 +27,13 @@ Version history
 * redis: performance improvement for larger split commands by avoiding string copies.
 * redis: correctly follow MOVE/ASK redirection for mirrored clusters.
 * redis: add :ref:`host_degraded_refresh_threshold <envoy_api_field_config.cluster.redis.RedisClusterConfig.host_degraded_refresh_threshold>` and :ref:`failure_refresh_threshold <envoy_api_field_config.cluster.redis.RedisClusterConfig.failure_refresh_threshold>` to refresh topology when nodes are degraded or when requests fails.
+* router: added histograms to show timeout budget usage to the :ref:`cluster stats <config_cluster_manager_cluster_stats>`.
 * router: added support for REQ(header-name) :ref:`header formatter <config_http_conn_man_headers_custom_request_headers>`.
 * router: allow using a :ref:`query parameter
   <envoy_api_field_route.RouteAction.HashPolicy.query_parameter>` for HTTP consistent hashing.
+* router: exposed DOWNSTREAM_REMOTE_ADDRESS as custom HTTP request/response headers.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: nit: I think you can revert this move, but nbd

File: source/common/upstream/upstream_impl.cc [hunk 15]
```diff
@@ -643,6 +647,14 @@ ClusterInfoImpl::ClusterInfoImpl(
       socket_matcher_(std::move(socket_matcher)), stats_scope_(std::move(stats_scope)),
       stats_(generateStats(*stats_scope_)), load_report_stats_store_(stats_scope_->symbolTable()),
       load_report_stats_(generateLoadReportStats(load_report_stats_store_)),
+      timeout_budget_stats_store_(
+          config.track_timeout_budgets()
+              ? std::make_unique<Stats::IsolatedStoreImpl>(stats_scope_->symbolTable())
+              : nullptr),
```
Comment (mattklein123) [inline source/common/upstream/upstream_impl.cc]: Why is this needed? I think you want to use the main scope/store for the cluster to add your stats?

File: source/common/upstream/upstream_impl.h [hunk 16]
```diff
@@ -605,6 +609,8 @@ class ClusterInfoImpl : public ClusterInfo, protected Logger::Loggable<Logger::I
   mutable ClusterStats stats_;
   Stats::IsolatedStoreImpl load_report_stats_store_;
   mutable ClusterLoadReportStats load_report_stats_;
+  const std::unique_ptr<Stats::IsolatedStoreImpl> timeout_budget_stats_store_;
+  mutable absl::optional<ClusterTimeoutBudgetStats> timeout_budget_stats_;
```
Comment (mattklein123) [inline source/common/upstream/upstream_impl.h]: Why does this need to be mutable?

File: source/common/router/router.cc [hunk 17]
```diff
@@ -1373,6 +1416,19 @@ Filter::UpstreamRequest::UpstreamRequest(Filter& parent, Http::ConnectionPool::I
     }
   }
 
+  if (parent_.cluster_->timeoutBudgetStats().has_value()) {
+    const auto start_time = parent_.callbacks_->dispatcher().timeSource().monotonicTime();
+    timeout_budget_stats_ = std::make_unique<Cleanup>([this, start_time]() {
```
Comment (mattklein123) [inline source/common/router/router.cc]: Instead of using a Cleanup, can you just send the stat when the per-try completes? From a readability perspective this seems a bit convoluted unless there is a good reason to do it.

File: source/common/router/router.cc [hunk 18]
```diff
@@ -804,6 +832,15 @@ void Filter::onPerTryTimeout(UpstreamRequest& upstream_request) {
...
   }
 
+  if (cluster_->timeoutBudgetStats().has_value()) {
+    Event::Dispatcher& dispatcher = callbacks_->dispatcher();
+    std::chrono::milliseconds response_time = std::chrono::duration_cast<std::chrono::milliseconds>(
+        dispatcher.timeSource().monotonicTime() - downstream_request_complete_time_);
+
+    cluster_->timeoutBudgetStats()->upstream_rq_timeout_budget_percent_used_.recordValue(
+        percentageOfTimeout(response_time, timeout_.global_timeout_));
+  }
```
Comment (mattklein123) [inline source/common/router/router.cc]: I'm confused. This is tracking the global timeout in the per-try timeout case?

File: source/common/router/router.cc [hunk 19]
```diff
@@ -739,10 +754,23 @@ void Filter::onResponseTimeout() {
...
+
+      if (cluster_->timeoutBudgetStats().has_value()) {
+        // Cancel firing per-try timeout information, because the per-try timeout did not come into
+        // play when the global timeout was hit.
+        upstream_request->timeout_budget_stats_->cancel();
+
+        Event::Dispatcher& dispatcher = callbacks_->dispatcher();
+        std::chrono::milliseconds response_time =
+            std::chrono::duration_cast<std::chrono::milliseconds>(
+                dispatcher.timeSource().monotonicTime() - downstream_request_complete_time_);
```
Comment (mattklein123) [inline source/common/router/router.cc]: In this case, you know the timeout as it was the global timeout, right? Can you just use it directly to be more clear?

File: source/common/upstream/upstream_impl.h (hunk 16 above)
Comment (gumpt) [reply source/common/upstream/upstream_impl.h]: Was cargo-culting against upstream.h to only define const pure virtual functions.  Will adjust this and the rest of the signatures.

Comment (gumpt) [reply source/common/upstream/upstream_impl.h]: Immediate results are mixed -- `const ClusterInfo` is passed around, going to play with it a bit.

File: source/common/router/router.cc (hunk 17 above)
Comment (gumpt) [reply source/common/router/router.cc]: It seemed more convenient to have it fire with the correct value whenever the try either times out or completes successfully without putting the same stat tracking code in a few different places.  This also lets the stat be cancelled if the global timeout hits first instead of recording a value into the per-try histogram when the per-try timeout didn't play into the request's timeout.  @tonya11en and I had a discussion about this offline and here: https://github.com/envoyproxy/envoy/pull/9227#discussion_r359545490

If it's less readable, I think I have to move this to a few different places.  I can try to figure out another way to preserve the behavior discussed above w/r/t the global timeout cancelling it.

If there's a difference of opinion on what should be in the histograms, I would also change this to address readability concerns.

Comment (mattklein123) [reply source/common/router/router.cc]: IMO this is a bit magical and I would prefer it be explicit, but I don't feel that strongly about it as long as it has good comments.

File: source/common/router/router.cc (hunk 18 above)
Comment (gumpt) [reply source/common/router/router.cc]: This was me messing up and putting this in the wrong place.  Will fix.

File: source/common/router/router.cc (hunk 19 above)
Comment (gumpt) [reply source/common/router/router.cc]: Yes, this should just fire 100.

**Comment by gumpt** [comment]
Commenting for this to be unmarked as stale, will push the buttons to re-request review as well.

**Comment by mattklein123** [comment]
Sorry do you mind merging master and then I will take another pass over this?

/wait

File: test/mocks/upstream/cluster_info.h [hunk 20]
```diff
@@ -130,6 +131,8 @@ class MockClusterInfo : public ClusterInfo {
   Upstream::TransportSocketMatcherPtr transport_socket_matcher_;
   NiceMock<Stats::MockIsolatedStatsStore> load_report_stats_store_;
   ClusterLoadReportStats load_report_stats_;
+  NiceMock<Stats::MockIsolatedStatsStore> timeout_budget_stats_store_;
```
Comment (gumpt) [inline test/mocks/upstream/cluster_info.h]: Delete this.

File: source/common/router/router.cc [hunk 21]
```diff
@@ -859,6 +880,15 @@ void Filter::chargeUpstreamAbort(Http::Code code, bool dropped, UpstreamRequest&
 
 void Filter::onUpstreamTimeoutAbort(StreamInfo::ResponseFlag response_flags,
                                     absl::string_view details) {
+  if (cluster_->timeoutBudgetStats().has_value()) {
+    Event::Dispatcher& dispatcher = callbacks_->dispatcher();
+    std::chrono::milliseconds response_time = std::chrono::duration_cast<std::chrono::milliseconds>(
+        dispatcher.timeSource().monotonicTime() - downstream_request_complete_time_);
+
+    cluster_->timeoutBudgetStats()->upstream_rq_timeout_budget_percent_used_.recordValue(
+        percentageOfTimeout(response_time, timeout_.global_timeout_));
```
Comment (mattklein123) [inline source/common/router/router.cc]: I may have asked this previously, but in this case don't you know that you know that you used the global_timeout_ so can't you just provide 100 as the used percent?

File: source/common/router/router.cc [hunk 22]
```diff
@@ -1381,6 +1417,23 @@ Filter::UpstreamRequest::UpstreamRequest(Filter& parent, Http::ConnectionPool::I
     }
   }
 
+  // If configured, set up the per-try histogram to fire when the UpstreamRequest
+  // completes. This is done via a Cleanup so that it can be cancelled if the global
+  // timeout is what cancels the request instead of the per-try timeout, and then avoid
+  // writing a misleading value into the histogram.
```
Comment (mattklein123) [inline source/common/router/router.cc]: I get why you are doing this, but are there really more cases other than per-try timeout and upstream complete? Can't we just call a utility function directly in those cases, with the per-try timeout case passing 100 per my comment above? It seems like if we don't need a Cleanup in the global case we don't need one here. The reason I keep coming back to this is this code is already incredibly complicated and magical stuff like this makes it even harder to grok the control flow.

File: source/common/router/router.cc (hunk 21 above)
Comment (gumpt) [reply source/common/router/router.cc]: I think I had agreed with you because this was in the path of the global timeout only.  Now this is in the codepath for either exhaustion of per-try requests or the global timeout failing.

In the case where all retries have failed and this is hit, this shouldn't be 100.

File: source/common/router/router.cc (hunk 22 above)
Comment (gumpt) [reply source/common/router/router.cc]: I think it would be cleaner for the global to use a Cleanup as well since it has two different places to fire mostly identical code.  It doesn't need the cancellable aspect of the Cleanup, so I put it in two places to avoid leaning on RAII magic.

I don't have better ideas for how to keep the per-try firing cancellable than the Cleanup's method of assigning to a function that's called during the per-try destructor.  Is it less magic to set a member std::function variable to `[](){}` rather than call a named member function named cancel?

File: test/mocks/upstream/cluster_info.h (hunk 20 above)
Comment (gumpt) [reply test/mocks/upstream/cluster_info.h]: Not going to delete this -- turns out it's way harder to assert on the store for the histograms when they share an underlying store!

File: source/common/router/router.cc (hunk 22 above)
Comment (gumpt) [reply source/common/router/router.cc]: Slept on it. I’ll change this today.

Comment (mattklein123) [reply source/common/router/router.cc]: Great, thanks. 

/wait

**Comment by gumpt** [comment]
Fully expect the stats integration tests to fail, just get to wait since I don't have a machine to reproduce the correct values.

**Comment by gumpt** [comment]
Going to re-merge master for the stats tests.

/wait

**Comment by gumpt** [comment]
Still waiting for new values.

/wait

**Comment by gumpt** [comment]
Going to merge master again.

**Comment by gumpt** [comment]
Test integration foobar

/wait

File: docs/root/intro/version_history.rst [hunk 23]
```diff
@@ -30,6 +30,8 @@ Version history
 * redis: performance improvement for larger split commands by avoiding string copies.
 * redis: correctly follow MOVE/ASK redirection for mirrored clusters.
 * redis: add :ref:`host_degraded_refresh_threshold <envoy_api_field_config.cluster.redis.RedisClusterConfig.host_degraded_refresh_threshold>` and :ref:`failure_refresh_threshold <envoy_api_field_config.cluster.redis.RedisClusterConfig.failure_refresh_threshold>` to refresh topology when nodes are degraded or when requests fails.
+* router: added histograms to show timeout budget usage to the :ref:`cluster stats <config_cluster_manager_cluster_stats>`.
+* router: added support for REQ(header-name) :ref:`header formatter <config_http_conn_man_headers_custom_request_headers>`.
```
Comment (gumpt) [inline docs/root/intro/version_history.rst]: Delete the duplicate note.

File: test/integration/stats_integration_test.cc [hunk 24]
```diff
@@ -265,11 +265,12 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithFakeSymbolTable) {
   // 2019/10/17  8484     43340       44000   stats: add unit support to histogram
   // 2019/11/01  8859     43563       44000   build: switch to libc++ by default
   // 2019/11/15  9040     43371       44000   build: update protobuf to 3.10.1
-  // 2019/11/15  9040     43403       44000   upstream: track whether cluster is local
+  // 2019/11/15  9031     43403       44000   upstream: track whether cluster is local
```
Comment (mattklein123) [inline test/integration/stats_integration_test.cc]: I think you have several merge issues in this file.

Comment (gumpt) [reply test/integration/stats_integration_test.cc]: The 9040->9031 is intentional, the PR did not update this line correctly.  #9040 was the build commit listed above it, #9031 is the one mentioned in the description.

Will fix the 4400->43500 mistake, looks like that value was changed sometime after the original commit.
